File: agent/component/wencai.py
#
#  Copyright 2024 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
from abc import ABC
import pandas as pd
import pywencai
from agent.component.base import ComponentBase, ComponentParamBase


class WenCaiParam(ComponentParamBase):
    """
    Define the WenCai component parameters.
    """

    def __init__(self):
        super().__init__()
        self.top_n = 10
        self.query_type = "stock"

    def check(self):
        self.check_positive_integer(self.top_n, "Top N")
        self.check_valid_value(self.query_type, "Query type",
                               ['stock', 'zhishu', 'fund', 'hkstock', 'usstock', 'threeboard', 'conbond', 'insurance',
                                'futures', 'lccp',
                                'foreign_exchange'])


class WenCai(ComponentBase, ABC):
    component_name = "WenCai"

    def _run(self, history, **kwargs):
        ans = self.get_input()
        ans = ",".join(ans["content"]) if "content" in ans else ""
        if not ans:
            return WenCai.be_output("")

        try:
            wencai_res = []
            res = pywencai.get(query=ans, query_type=self._param.query_type, perpage=self._param.top_n)
            if isinstance(res, pd.DataFrame):
                wencai_res.append({"content": res.to_markdown()})
            if isinstance(res, dict):
                for item in res.items():
                    if isinstance(item[1], list):
                        wencai_res.append({"content": item[0] + "\n" + pd.DataFrame(item[1]).to_markdown()})
                        continue
                    if isinstance(item[1], str):
                        wencai_res.append({"content": item[0] + "\n" + item[1]})
                        continue
                    if isinstance(item[1], dict):
                        if "meta" in item[1].keys():
                            continue
                        wencai_res.append({"content": pd.DataFrame.from_dict(item[1], orient='index').to_markdown()})
                        continue
                    if isinstance(item[1], pd.DataFrame):
                        if "image_url" in item[1].columns:
                            continue
                        wencai_res.append({"content": item[1].to_markdown()})
                        continue
                        
                    wencai_res.append({"content": item[0] + "\n" + str(item[1])})
        except Exception as e:
            return WenCai.be_output("**ERROR**: " + str(e))

        if not wencai_res:
            return WenCai.be_output("")

        return pd.DataFrame(wencai_res)

File: internal/agent/component/wencai.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package component

import (
	"context"
	"encoding/json"
	"fmt"

	agenttool "ragflow/internal/agent/tool"
)

type wencaiInvoker interface {
	InvokableRun(ctx context.Context, argsJSON string) (string, error)
}

type wencaiComponent struct {
	inner wencaiInvoker
}

func newWencaiComponent(params map[string]any) (Component, error) {
	toolParams := make(map[string]any, 2)
	for _, key := range []string{"top_n", "query_type"} {
		if value, ok := params[key]; ok {
			toolParams[key] = value
		}
	}
	inner, err := agenttool.BuildByName("wencai", toolParams)
	if err != nil {
		return nil, err
	}
	invoker, ok := inner.(wencaiInvoker)
	if !ok {
		return nil, fmt.Errorf("WenCai: tool does not implement InvokableRun")
	}
	return newWencaiComponentWithInvoker(invoker), nil
}

func newWencaiComponentWithInvoker(inner wencaiInvoker) Component {
	return &wencaiComponent{inner: inner}
}

func (c *wencaiComponent) Name() string { return "WenCai" }

func (c *wencaiComponent) Inputs() map[string]string {
	return map[string]string{
		"query": "The question/conditions to select stocks.",
	}
}

func (c *wencaiComponent) Outputs() map[string]string {
	return map[string]string{
		"report": "WenCai query report.",
	}
}

func (c *wencaiComponent) GetInputForm() map[string]any {
	return map[string]any{
		"query": map[string]any{
			"name": "Query",
			"type": "line",
		},
	}
}

func (c *wencaiComponent) Invoke(ctx context.Context, inputs map[string]any) (map[string]any, error) {
	query := stringParam(inputs["query"])
	if query == "" {
		return map[string]any{"report": ""}, nil
	}
	argsJSON, err := json.Marshal(map[string]any{"query": query})
	if err != nil {
		return nil, fmt.Errorf("canvas: WenCai: encode query: %w", err)
	}
	out, err := c.inner.InvokableRun(ctx, string(argsJSON))
	decoded := parseToolEnvelope(out)
	report, _ := decoded["report"].(string)
	if message, _ := decoded["_ERROR"].(string); message != "" {
		return map[string]any{"report": report, "_ERROR": message}, nil
	}
	if err != nil {
		return nil, fmt.Errorf("canvas: WenCai: %w", err)
	}
	return map[string]any{"report": report}, nil
}

func (c *wencaiComponent) Stream(_ context.Context, _ map[string]any) (<-chan map[string]any, error) {
	return nil, nil
}

func init() {
	Register("WenCai", newWencaiComponent)
}
